service/storage_proxy.cc

```diff
@@ -2155,36 +2156,45 @@ future<result<>> storage_proxy::mutate_begin(unique_response_handler_vector ids,
 
 // this function should be called with a future that holds result of mutation attempt (usually
 // future returned by mutate_begin()). The future should be ready when function is called.
-future<> storage_proxy::mutate_end(future<result<>> mutate_result, utils::latency_counter lc, write_stats& stats, tracing::trace_state_ptr trace_state) {
+future<result<>> storage_proxy::mutate_end(future<result<>> mutate_result, utils::latency_counter lc, write_stats& stats, tracing::trace_state_ptr trace_state) {
     assert(mutate_result.available());
     stats.write.mark(lc.stop().latency());
     if (lc.is_start()) {
         stats.estimated_write.add(lc.latency());
     }
+    auto exception_handler = overloaded_functor {
+        [&] (const mutation_write_timeout_exception& ex) {
+            // timeout
+            tracing::trace(trace_state, "Mutation failed: write timeout; received {:d} of {:d} required replies", ex.received, ex.block_for);
+            slogger.debug("Write timeout; received {} of {} required replies", ex.received, ex.block_for);
+            stats.write_timeouts.mark();
+        }
+    };
     try {
-        mutate_result.get().value(); // .value() throws the error, if there is any
+        auto res = mutate_result.get();
+        if (res.has_error()) {
+            res.assume_error().accept(exception_handler);
+            return make_ready_future<result<>>(std::move(res));
+        }
         tracing::trace(trace_state, "Mutation successfully completed");
-        return make_ready_future<>();
+        return make_ready_future<result<>>(bo::success());
     } catch (replica::no_such_keyspace& ex) {
         tracing::trace(trace_state, "Mutation failed: write to non existing keyspace: {}", ex.what());
         slogger.trace("Write to non existing keyspace: {}", ex.what());
-        return make_exception_future<>(std::current_exception());
+        return make_exception_future<result<>>(std::current_exception());
     } catch(mutation_write_timeout_exception& ex) {
-        // timeout
-        tracing::trace(trace_state, "Mutation failed: write timeout; received {:d} of {:d} required replies", ex.received, ex.block_for);
-        slogger.debug("Write timeout; received {} of {} required replies", ex.received, ex.block_for);
-        stats.write_timeouts.mark();
-        return make_exception_future<>(std::current_exception());
+        exception_handler(ex);
+        return make_exception_future<result<>>(std::current_exception());
```

Comment:
The new way of returning exceptions is meant to allow gradual migration. If somebody decides to migrate another exception type, they don't have to modify all throw sites at once - it is easier to focus on the most common cases first and leave others as throws. You also don't have to migrate the whole exception handling path at once, and you can add .then(result_into_future<...>) to convert a future<result<>> to future<> (which can generate an exceptional future).

It's just safer to cover both exceptions and results in the same place. Even if, after this PR, this function will never receive mutation_write_timeout_exception as a regular exception, it will protect us in the future if somebody starts throwing it or it appears due to some result -> exception future conversion logic. I wanted to set a precedent.

Reply:
Ok, while I understand this thinking, the downside is that this "precedent" means that exception-handling code now always (?) needs to be duplicated twice, because supposedly you never know if the exception will come from a result or an actual exception.

We had exactly the same problem with futures - a future-returning function can either return an exceptional future or really throw - and our solution was in places like then(), futurize(), etc., to catch exceptions and turn them into exceptional futures. Maybe here as well, the code could just catch the exception and turn it into a result, without any duplicated logic. Maybe this is what you were doing and I just misunderstood?

Reply:
> Ok, while I understand this thinking, the downside is that this "precedent" means that exception-handling code now always (?) needs to be duplicated twice, because supposedly you never know if the exception will come from a result or an actual exception.

Well, the pattern I used requires some code duplication, yes. Given exception type, you need to both handle it in result and write a catch() clause, so that's adding code to two places - and I can agree that it is fragile. The core of the exception handler is not duplicated as I moved it to the overloaded_functor and the catch() clause just calls it.

> We had exactly the same problem with futures - a future-returning function can either return an exceptional future or really throw - and our solution was in places like then(), futurize(), etc., to catch exceptions and turn them into exceptional futures. Maybe here as well, the code could just catch the exception and turn it into a result, without any duplicated logic. Maybe this is what you were doing and I just misunderstood?

I don't think it is possible to turn an exceptional future into a failed result<>. You can't, in general, take out an exception out of std::exception_ptr. If you rethrow it and do `catch (const std::exception& ex)`, the standard says that the reference is valid as long as the error handler is executing, so you can't return it from a function. On the other hand, you can't move out or copy the exception because it can be a derived class and you would have to somehow be able to dynamically invoke the correct copy constructor. Not to mention exceptions which are not derived from std::exception (we have catch (...) for that in the current logic).

I think I have an idea for a metaprogramming-powered way to handle results<>, exceptions and exceptional futures in a unified way. It would be used like this:

```c++
return utils::try_catch(
    [&] () { return some_operation_returning_future_result(); },
    result_catch<exception1>([&] (const auto& ex, auto&& rethrow) {
        return make_ready_future<result<>>(bo::success());
    }),
    result_catch<exception2>([&] (const auto& ex, auto&& rethrow) {
        // Handle the exception here
        // Because we don't want to convert between results and exceptions,
        // if we want to rethrow then we need to return different things.
        // Hence, the generic rethrow argument.
        return rethrow();
    }),
    result_catch_all([&] (auto&& rethrow) {
        // Equivalent to catch (...)
        return rethrow();
    })
);
```

I'm not sure if it won't be an overkill, though.

Reply:
I actually managed to write the `try_catch` function and it both looks and works nicely. However, it is definitely more complicated than the other result-related helper function and the change introducing it is not that small, so I would like to postpone it to a followup PR so that it can be reviewed independently.